src/lcf/string_view.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * This file is part of liblcf. Copyright (c) 2020 liblcf authors.
+ * https://github.com/EasyRPG/liblcf - https://easyrpg.org
+ *
+ * liblcf is Free/Libre Open Source Software, released under the MIT License.
+ * For the full copyright and license information, please view the COPYING
+ * file that was distributed with this source code.
+ */
+
+#ifndef LCF_STRING_VIEW_H
+#define LCF_STRING_VIEW_H
+#include <cstring>
+#include <string>
+#include <cassert>
+#include <iterator>
+#include <ostream>
+#include <algorithm>
+
+#define nssv_CONFIG_NO_EXCEPTIONS 1
+#define nssv_CONFIG_CONVERSION_STD_STRING 1
+#include <lcf/third_party/string_view.h>
+
+namespace lcf {
+
+template <typename CharT, typename Traits = std::char_traits<CharT>>
+	using BasicStringView = nonstd::basic_string_view<CharT,Traits>;
+
+using StringView = BasicStringView<char>;
+using WStringView = BasicStringView<wchar_t>;
+using U16StringView = BasicStringView<char16_t>;
+using U32StringView = BasicStringView<char32_t>;
+
+template< class CharT, class Traits, class Allocator = std::allocator<CharT> >
+std::basic_string<CharT, Traits, Allocator>
+ToString(BasicStringView<CharT,Traits> sv, const Allocator& a = Allocator()) {
+	return nonstd::to_string(sv, a);
+}
+
+template< class CharT, class Traits, class Allocator >
+BasicStringView<CharT, Traits>
+ToStringView(const std::basic_string<CharT, Traits, Allocator>& s )
+{
+	return nonstd::to_string_view(s);
+}
+
+// Make sure converting string to string is an error. To catch bugs.
+template <typename C, typename T, typename A>
+void ToString(const std::basic_string<C,T,A>&) = delete;
+
+// We use macros instead of template to enable implicit string -> string_view conversion here.
```

Comment:
Can you give an example what you get with these macros that looks much worse without?

Reply:
With the macro I can write this:
```
std::string x;
Substr(x, 2, 3);
```

Without it, I have to write this
```
std::string x;
Substr(ToStringView(x), 2, 3);
```

Converting string -> string_view is normal and doesn't cost in performance, so it should be allowed freely. Going the other way means making a copy and thus should be flagged.

```
std::string y;
StringView x;
x = y; // Ok
y = x; // Error
y = ToString(x); // Ok, you meant to make a copy here.
```